java/core/src/test/java/org/bondlib/StructGenerator.java

```diff
@@ -0,0 +1,408 @@
+            returnValue = this.generateStruct(structBondType, context);
+        } else {
+            // unknown
+            throw new IllegalArgumentException("Unknown BondType implementation: " + bondTypeClass);
+        }
+        return returnValue;
+    }
+
+    // returns true if a pseudo-random number distributed uniformly between 0 and 1 is less than the argument
+    private boolean testProbability(double probability) {
+        return this.rand.nextDouble() < probability;
+    }
+
+    // helper to generate a pseudo-random integer with the given number of bits (optionally signed)
+    private long generateIntValue(int bitCount, boolean isSigned) {
+        // Given number N randomly chosen between 1 and bitCount with uniform probability, determine the
+        // magnitude of the result as an unsigned value of N bits where bits 0 to N-1 are randomly chosen
+        // individually (with uniform probability), and all the other bits are set to 0. Then optionally
+        // negate the result with 50% probability.
+        // This algorithm improves coverage of smaller integers for variable-length encoding.
+        int n = this.rand.nextInt(bitCount) + 1;
+        long result = 0;
+        for (int i = 0; i < n; ++i) {
+            if (this.rand.nextBoolean()) {
+                result |= (1 << i);
+            }
+        }
+        if (isSigned && this.rand.nextBoolean()) {
+            result = -result;
+        }
+        return result;
+    }
+
+    // helper to generate a pseudo-random string
+    private String generateStringValue() {
+        int length = this.rand.nextInt(this.settings.maxGeneratedStringLength + 1);
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < length; ++i) {
+            // pick the character (as Unicode code point) randomly
+            int codePointIndex = rand.nextInt(validUnicodeCodePoints.size());
+            int codePoint = validUnicodeCodePoints.get(codePointIndex);
+            sb.appendCodePoint(codePoint);
+        }
+        return sb.toString();
+    }
+
+    // helper to generate a pseudo-random enum value
+    private BondEnum generateEnumValue(EnumBondType<?> enumBondType) {
+        // use reflection to invoke the static member to create an enum value
+        try {
+            Method getMethod = enumBondType.getValueClass().getMethod("get", int.class);
+            int int32Value = (int) this.generateIntValue(31, true);
+            return (BondEnum) getMethod.invoke(null, int32Value);
+        } catch (Exception ex) {
+            // shouldn't happen
+            throw new RuntimeException(ex);
+        }
+    }
+
+    // helper to generate collection (vector/list/set) elements
+    @SuppressWarnings("unchecked")
+    private void generateCollectionElements(
+            Collection collection, BondType elementType, GenerationContext context) {
+        if (!context.hasReachedStructRecursionDepthLimit(elementType, context)) {
```

Comment:
The idea is that if you're make a `T` inside of a collection instead of as a member, it still counts for depth limits? This doesn't increment the recursion because it's counting on `generateValue` to do that?

Reply:
generateValue is general and will call generateStruct if the element is a struct, which will increment depth.

---
In reply to: [138755007](https://github.com/Microsoft/bond/pull/607#discussion_r138755007) [](ancestors = 138755007)